### ISSUE TITLE
snapdev=visible should be default

### ISSUE THREAD

**Issue by brendonhumphrey** [openzfsonosx/zfs#486]
Based on test scripts and FreeBSD behaviour it seems that zvol snapshots should be visible by default - as root:

zpool create tank da1
zfs create tank/testvol

ls -l /dev/zvol/tank/testvol*
.... /dev/zvol/tank/testvol

zfs snapshot tank/testvol@1

ls -l /dev/zvol/tank/testvol*
.... /dev/zvol/tank/testvol
.... /dev/zvol/tank/testvol@1

We appear to comply with this provided snapdev=visible on the zvol.

**Comment by evansus** [comment]
@brendonhumphrey this is true, but the behavior on OS X of snapdev=visible is pretty rough for end-users, even if we prevent automount on zvols and/or snapshots.

Every time a snapshot is made, a read-only zvol device is presented to the OS. If you do periodic daily or hourly snapshots of an entire pool which contains multiple zvols, that means several zvols appearing and possibly being mounted each time.

For example, I had a large HFS zvol for Time Machine, and every hour the snapshot would be presented. On the desktop were multiple "Time Machine" volumes, and after a reboot it would fail to backup as it would get confused and try to use a read-only snapshot.

I believe the recent changes to zvol_create_minors have resolved this last part, by creating a minor for the zvol prior to the snapshots. Even if so, a read-only snapshot may mount sooner than it's read-write parent, since the mount process is more involved for a read-write volume. With matching UUIDs and mounting 'out of order', Time Machine doesn't know which to use. It would be nice, but presumptuous, to use the read-write volume and ignore the read-only ones - Apple errs on the side of caution.

IMO it would be ideal to leave automount on and snapdev=hidden, but at the very least we need to allow users to control both. Also IMO, but as discussed with @lundman and @ilovezfs in the past, zvol snapshots have little positive impact on Mac OS X, mostly negatives, and any use-case for them can be achieved by the user by either doing `zfs clone` on an individual snapshot, or `zfs set snapdev=visible` on the zvol and its snapshots as needed.
